### PR TITLE
PSA: Phase 3: x509: use the PSA API to perform hashing operations

### PR DESCRIPTION
**Pull request by AndrzejKurek**
So far limited only to certificate verification withour CRL and CSR generation.

This PR resolves iotssl-2597.
The helper functions are to be moved to a separate file.
This PR supersedes https://github.com/ARMmbed/mbedtls-psa/pull/180.

### CODE REVIEW

File: library/x509_crt.c [hunk 1]
```diff
@@ -1893,16 +1940,36 @@ static int x509_crt_check_signature( const mbedtls_x509_crt *child,
                                      mbedtls_x509_crt *parent,
                                      mbedtls_x509_crt_restart_ctx *rs_ctx )
 {
-    const mbedtls_md_info_t *md_info;
     unsigned char hash[MBEDTLS_MD_MAX_SIZE];
-
+    size_t hash_len;
+    /* Note: hash errors can happen only after an internal error */
```
Comment (hanno-arm) [inline library/x509_crt.c]: This comment is slightly misplaced now; I'd prefer to remove it entirely or move it L.1950 again.

File: library/x509write_csr.c [hunk 2]
```diff
@@ -187,9 +239,25 @@ int mbedtls_x509write_csr_der( mbedtls_x509write_csr *ctx, unsigned char *buf, s
...
+    if( psa_hash_setup( &hash_operation, hash_alg ) != PSA_SUCCESS )
+        return( MBEDTLS_ERR_X509_FATAL_ERROR );
 
+    if( psa_hash_update( &hash_operation, c, len) != PSA_SUCCESS )
+    {
+        psa_hash_abort( &hash_operation );
+        return( MBEDTLS_ERR_X509_FATAL_ERROR );
+    }
+    if( psa_hash_finish( &hash_operation, hash, sizeof( hash ), &hash_len )
+        != PSA_SUCCESS) {
```
Comment (hanno-arm) [inline library/x509write_csr.c]: Minor: Please add a line break before `{`.

File: library/x509write_csr.c [hunk 3]
```diff
@@ -136,7 +184,11 @@ int mbedtls_x509write_csr_der( mbedtls_x509write_csr *ctx, unsigned char *buf, s
     size_t pub_len = 0, sig_and_oid_len = 0, sig_len;
     size_t len = 0;
     mbedtls_pk_type_t pk_alg;
-
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    psa_hash_operation_t hash_operation;
+    size_t hash_len;
+    psa_algorithm_t hash_alg = translate_md_to_psa(ctx->md_alg);
```
Comment (hanno-arm) [inline library/x509write_csr.c]: Minor: Add spaces.

Comment (AndrzejKurek) [reply library/x509write_csr.c]: Added.

File: library/x509write_csr.c (hunk 2 above)
Comment (AndrzejKurek) [reply library/x509write_csr.c]: Added.

File: library/x509_crt.c (hunk 1 above)
Comment (AndrzejKurek) [reply library/x509_crt.c]: Adjusted.

File: library/x509_crt.c [hunk 4]
```diff
@@ -1885,6 +1889,49 @@ static int x509_name_cmp( const mbedtls_x509_name *a, const mbedtls_x509_name *b
     /* a == NULL == b */
     return( 0 );
 }
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+static psa_algorithm_t translate_md_to_psa( mbedtls_md_type_t md_alg )
```
Comment (mpg) [inline library/x509_crt.c]: This will be removed when rebasing over #2184

File: library/x509write_csr.c [hunk 5]
```diff
@@ -122,6 +126,50 @@ int mbedtls_x509write_csr_set_ns_cert_type( mbedtls_x509write_csr *ctx,
     return( 0 );
 }
 
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+static psa_algorithm_t translate_md_to_psa( mbedtls_md_type_t md_alg )
```
Comment (mpg) [inline library/x509write_csr.c]: Same as above.

File: library/x509_crt.c [hunk 6]
```diff
@@ -1893,16 +1898,37 @@ static int x509_crt_check_signature( const mbedtls_x509_crt *child,
...
+    
+    /* Note: hash errors can happen only after an internal error */
     if( mbedtls_md( md_info, child->tbs.p, child->tbs.len, hash ) != 0 )
+        return( -1 );
+#else
+    psa_hash_operation_t hash_operation;
+    psa_algorithm_t hash_alg = mbedtls_psa_translate_md( child->sig_md );
+
+    if( psa_hash_setup( &hash_operation, hash_alg ) != PSA_SUCCESS )
+        return( -1 );
```
Comment (hanno-arm) [inline library/x509_crt.c]: Can we return `MBEDTLS_ERR_PK_HW_ACCEL_FAILED` here?

File: library/x509_crt.c [hunk 7]
```diff
@@ -1893,16 +1898,37 @@ static int x509_crt_check_signature( const mbedtls_x509_crt *child,
...
+
+    if( psa_hash_update( &hash_operation, child->tbs.p, child->tbs.len )
+        != PSA_SUCCESS )
     {
-        /* Note: this can't happen except after an internal error */
+        psa_hash_abort( &hash_operation );
```
Comment (hanno-arm) [inline library/x509_crt.c]: According to https://github.com/ARMmbed/mbedtls/blob/feature-psa/include/psa/crypto.h#L2015, we don't have to call `psa_hash_abort()` here.

File: library/x509_crt.c [hunk 8]
```diff
@@ -1893,16 +1898,37 @@ static int x509_crt_check_signature( const mbedtls_x509_crt *child,
...
         return( -1 );
     }
 
+    if( psa_hash_finish( &hash_operation, hash, sizeof( hash ), &hash_len )
+        != PSA_SUCCESS )
+    {
+        psa_hash_abort( &hash_operation );
```
Comment (hanno-arm) [inline library/x509_crt.c]: Same here.

File: library/x509write_csr.c [hunk 9]
```diff
@@ -187,9 +196,26 @@ int mbedtls_x509write_csr_der( mbedtls_x509write_csr *ctx, unsigned char *buf, s
...
+     * Note: hash errors can happen only after an internal error
      */
-    mbedtls_md( mbedtls_md_info_from_type( ctx->md_alg ), c, len, hash );
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( psa_hash_setup( &hash_operation, hash_alg ) != PSA_SUCCESS )
+        return( MBEDTLS_ERR_X509_FATAL_ERROR );
 
+    if( psa_hash_update( &hash_operation, c, len) != PSA_SUCCESS )
+    {
+        psa_hash_abort( &hash_operation );
```
Comment (hanno-arm) [inline library/x509write_csr.c]: Same here.

File: library/x509write_csr.c [hunk 10]
```diff
@@ -187,9 +196,26 @@ int mbedtls_x509write_csr_der( mbedtls_x509write_csr *ctx, unsigned char *buf, s
...
+        return( MBEDTLS_ERR_X509_FATAL_ERROR );
+    }
+    if( psa_hash_finish( &hash_operation, hash, sizeof( hash ), &hash_len )
+        != PSA_SUCCESS )
+    {
+        psa_hash_abort( &hash_operation );
```
Comment (hanno-arm) [inline library/x509write_csr.c]: Same here.

File: library/x509write_csr.c [hunk 11]
```diff
@@ -187,9 +196,26 @@ int mbedtls_x509write_csr_der( mbedtls_x509write_csr *ctx, unsigned char *buf, s
...
     /*
      * Prepare signature
+     * Note: hash errors can happen only after an internal error
      */
-    mbedtls_md( mbedtls_md_info_from_type( ctx->md_alg ), c, len, hash );
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( psa_hash_setup( &hash_operation, hash_alg ) != PSA_SUCCESS )
+        return( MBEDTLS_ERR_X509_FATAL_ERROR );
 
+    if( psa_hash_update( &hash_operation, c, len) != PSA_SUCCESS )
```
Comment (hanno-arm) [inline library/x509write_csr.c]: Minor: Space missing after `len`.

**Comment by AndrzejKurek** [comment]
@mpg, @hanno-arm - I've addressed all of your comments and rebased the PR on psa-integration-utilities. Please re-review.

**Pull request merged**